sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/FrameworkScheduler.java

```diff
@@ -0,0 +1,269 @@
+        this.frameworkStore = frameworkStore;
+        this.abstractScheduler = abstractScheduler;
+        this.offerProcessor = offerProcessor;
+        this.stateStore = stateStore;
+    }
+
+    /**
+     * Notifies this instance that the API server has been initialized. All offers are declined until this is called.
+     *
+     * @return {@code this}
+     */
+    public FrameworkScheduler setReadyToAcceptOffers() {
+        readyToAcceptOffers.set(true);
+        return this;
+    }
+
+    /**
+     * Disables multithreading for tests. For this to take effect, it must be invoked before the framework has
+     * registered.
+     *
+     * @return {@code this}
+     */
+    @VisibleForTesting
+    public FrameworkScheduler disableThreading() {
+        offerProcessor.disableThreading();
+        this.multithreaded = false;
+        return this;
+    }
+
+    @Override
+    public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, Protos.MasterInfo masterInfo) {
+        if (isRegisterStarted.getAndSet(true)) {
+            // This may occur as the result of a master election.
+            LOGGER.info("Already registered, calling reregistered()");
+            reregistered(driver, masterInfo);
+            return;
+        }
+
+        Driver.setDriver(driver);
+        abstractScheduler.registeredWithMesos();
+
+        LOGGER.info("Registered framework with frameworkId: {}", frameworkId.getValue());
+        this.taskCleaner = new TaskCleaner(stateStore, multithreaded);
+
+        try {
+            frameworkStore.storeFrameworkId(frameworkId);
+        } catch (Exception e) {
+            LOGGER.error(String.format(
+                    "Unable to store registered framework ID '%s'", frameworkId.getValue()), e);
+            SchedulerUtils.hardExit(SchedulerErrorCode.REGISTRATION_FAILURE);
+        }
+
+        postRegister(driver, masterInfo);
+
+        offerProcessor.start();
+    }
+
+    @Override
+    public void reregistered(SchedulerDriver driver, Protos.MasterInfo masterInfo) {
+        LOGGER.info("Re-registered with master: {}", TextFormat.shortDebugString(masterInfo));
+        Driver.setDriver(driver);
+        postRegister(driver, masterInfo);
+    }
+
+    private void postRegister(SchedulerDriver driver, Protos.MasterInfo masterInfo) {
+        // Task reconciliation should be (re)started on all (re-)registrations.
+        Reconciler reconciler = offerProcessor.getReconciler();
+        reconciler.start();
+        reconciler.reconcile();
+        if (masterInfo.hasDomain()) {
+            IsLocalRegionRule.setLocalDomain(masterInfo.getDomain());
+        }
+    }
+
+    @Override
+    public void resourceOffers(SchedulerDriver driver, List<Protos.Offer> offers) {
+        Metrics.incrementReceivedOffers(offers.size());
+
+        if (!readyToAcceptOffers.get()) {
+            LOGGER.info("Declining {} offer{}: Waiting for API Server to start.",
```

Comment:
This log message seems off.  I think `readyToAcceptOffers` might be false for more reasons than just the API server not being up.  If not we should change the variable name.

Reply:
In practice this is only for the API server. Renamed `readyToAcceptOffers` to `apiServerStarted`, and `setReadyToAcceptOffers()` to `setApiServerStarted()`.